contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,352 @@
+
+        withdrawId = nextWithdrawId++;
+
+        withdrawals[withdrawId] = WithdrawRequest(
+            validatorId,
+            undelegateAmount,
+            block.timestamp
+        );
+        pendingWithdrawals += undelegateAmount;
+
+        sfc.undelegate(validatorId, withdrawId, undelegateAmount);
+
+        emit Undelegated(withdrawId, validatorId, undelegateAmount);
+    }
+
+    /**
+     * @notice Withdraw native S from a previously undelegated validator.
+     * The native S is wrapped wS and transferred to the Vault.
+     * @param withdrawId The withdraw ID returned from `undelegate` and emitted in `Undelegated`.
+     * @return withdrawnAmount The amount of Sonic (S) withdrawn.
+     * This can be less than the undelegated amount in the event of slashing.
+     */
+    // slither-disable-start reentrancy-no-eth
+    function withdrawFromSFC(uint256 withdrawId)
+        external
+        onlyRegistrator
+        nonReentrant
+        returns (uint256 withdrawnAmount)
+    {
+        require(withdrawId < nextWithdrawId, "Invalid withdrawId");
+
+        // Can still withdraw even if the validator is no longer supported
+        // Load the withdrawal from storage into memory
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(!isWithdrawnFromSFC(withdrawId), "Already withdrawn");
+
+        uint256 sBalanceBefore = address(this).balance;
+
+        sfc.withdraw(withdrawal.validatorId, withdrawId);
+
+        // Save state to storage
+        withdrawnAmount = address(this).balance - sBalanceBefore;
+        pendingWithdrawals -= withdrawal.undelegatedAmount;
+        withdrawals[withdrawId].undelegatedAmount = 0;
+
+        // Wrap Sonic (S) to Wrapped Sonic (wS)
+        IWrappedSonic(wrappedSonic).deposit{ value: withdrawnAmount }();
+
+        // Transfer the Wrapped Sonic (wS) to the Vault
+        // slither-disable-next-line unchecked-transfer unused-return
+        IERC20(wrappedSonic).transfer(vaultAddress, withdrawnAmount);
+
+        // withdrawal.undelegatedAmount & withdrawnAmount can differ in case of slashing
+        emit Withdrawn(
+            withdrawId,
+            withdrawal.validatorId,
+            withdrawal.undelegatedAmount,
+            withdrawnAmount
+        );
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /// @notice returns a bool whether a withdrawalId has already been withdrawn or not
+    function isWithdrawnFromSFC(uint256 withdrawId) public view returns (bool) {
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(withdrawal.validatorId > 0, "Invalid withdrawId");
+        return withdrawal.undelegatedAmount == 0;
+    }
+
+    /**
+     * @notice Restake any pending validator rewards for all supported validators
+     * @param validatorIds List of Sonic validator IDs to restake rewards
+     */
+    function restakeRewards(uint256[] calldata validatorIds)
```

Comment:
❓Do we need a way to collect rewards?

The SFC contract stores our funds in three buckets: Staked, stashed rewards, and calculated rewards. However, externally these can be treated as staked funds and reward funds. We have a mechanism to withdraw staked funds, but we do not have a mechanism to get at reward funds. This means that our only option to close out  delegator is to undelegate, wait, withdraw the staked funds, wait a while more, retake the rearwards to staked, and kick off a second undelegation cycle. This seems painful when we could just get our reward money right now.

Reply:
This is a great point thanks. Corrected here: https://github.com/OriginProtocol/origin-dollar/pull/2335/commits/129ce7fd78243befbaa874f62b96af9fd4b806c4